1_The_Physical/Level_2_Stateful_Guess_Point/Shared/Student/src/domain/core/aggregate-root.ts

```diff
@@ -0,0 +1,7 @@
+import { DomainEvent } from "./domain-event";
+
+export interface AggregateRoot<T> {
+  readonly id: string;
+  readonly state: T;
+  readonly events: DomainEvent[];
```

Comment:
I think this is fine for this task to have `events` in the aggregate root but consider the following.

First of all, what's the point to keep events and not use them? I assume, that it makes sense to utilize them to implement some kind of event sourcing (probably?). This means, that the state of `Student` should be hydrated from the events.

If you consider the above a way to go, then it adds some constraints:

1. Stored events should contain only essential state-changing data to hydrate the final state
2. The aggregate root might not be a good place to store the events. This is because if the implementation evolves then you might probably have Commands and Queries. An example of a query would be to fetch all students. To implement a such query within the current implementation you would need to create aggregate roots for all of the matching students, which might add a lot of overhead. If you decoupled storing of events from the aggregate root then it would simplify queries. Queries don't do any state changes so you don't need to enforce invariants for them, therefore aggregates and value objects are not necessary.

Reply:
I don't think it's a problem. I just wanted to attract your attention to the fact that just storing events without a purpose does not make any sense. If you keep in mind a possible purpose (it shouldn't be exactly what I mentioned) even without implementing it then it should positively affect the final solution.